x-pack/plugins/observability_solution/apm/server/lib/apm_telemetry/collect_data_telemetry/tasks.ts

```diff
@@ -898,193 +969,353 @@ export const tasks: TelemetryTask[] = [
-                .slice(0, size),
-              version: aggregations[AGENT_VERSION].buckets.map((bucket) => bucket.key as string),
-            },
-            service: {
-              framework: {
-                name: aggregations[SERVICE_FRAMEWORK_NAME].buckets
-                  .map((bucket) => bucket.key as string)
-                  .slice(0, size),
-                version: aggregations[SERVICE_FRAMEWORK_VERSION].buckets
-                  .map((bucket) => bucket.key as string)
-                  .slice(0, size),
-                composite: sortBy(
-                  flatten(
-                    aggregations[SERVICE_FRAMEWORK_NAME].buckets.map((bucket) =>
-                      bucket[SERVICE_FRAMEWORK_VERSION].buckets.map((versionBucket) => ({
-                        doc_count: versionBucket.doc_count,
-                        name: toComposite(bucket.key, versionBucket.key),
-                      }))
-                    )
-                  ),
-                  'doc_count'
-                )
-                  .reverse()
-                  .slice(0, size)
-                  .map((composite) => composite.name),
+          const initAgentData = OPEN_TELEMETRY_AGENT_NAMES.reduce((acc, agent) => {
+            return {
+              ...acc,
+              [agent]: {
+                agent: {
+                  activation_method: [],
+                  version: [],
+                },
+                service: {
+                  framework: {
+                    name: [],
+                    version: [],
+                    composite: [],
+                  },
+                  language: {
+                    name: [],
+                    version: [],
+                    composite: [],
+                  },
+                  runtime: {
+                    name: [],
+                    version: [],
+                    composite: [],
+                  },
+                },
               },
-              language: {
-                name: aggregations[SERVICE_LANGUAGE_NAME].buckets
-                  .map((bucket) => bucket.key as string)
-                  .slice(0, size),
-                version: aggregations[SERVICE_LANGUAGE_VERSION].buckets
-                  .map((bucket) => bucket.key as string)
-                  .slice(0, size),
-                composite: sortBy(
-                  flatten(
-                    aggregations[SERVICE_LANGUAGE_NAME].buckets.map((bucket) =>
-                      bucket[SERVICE_LANGUAGE_VERSION].buckets.map((versionBucket) => ({
-                        doc_count: versionBucket.doc_count,
-                        name: toComposite(bucket.key, versionBucket.key),
-                      }))
-                    )
+            };
+          }, {} as NonNullable<APMTelemetry['agents']>);
+
+          const agentData = aggregations?.agent_name.buckets.reduce((acc, agentNamesAggs) => {
+            return {
+              ...acc,
+              [agentName]: {
```

Comment:
Since we run two ES queries, the first one for known agent names and the second for generic otel agent names (`otlp/*` and `opentelemetry/*`) the part of the code responsible for formatting response data become duplicated. I tried to separate it but I run into typing issue. The problem is that `agentNamesAggs` type is inferred from the `telemetryClient.search` response, therefore it isn't known ahead of time to define it in the function.

this is duplicated part:
```ts
{
  agent: {
    activation_method: aggregations[AGENT_ACTIVATION_METHOD].buckets
      .map((bucket) => bucket.key as string)
  .slice(0, size),
      version: aggregations[AGENT_VERSION].buckets.map((bucket) => bucket.key as string),
  },
  service: {
    framework: {
      name: aggregations[SERVICE_FRAMEWORK_NAME].buckets
        .map((bucket) => bucket.key as string)
    .slice(0, size),
        version: aggregations[SERVICE_FRAMEWORK_VERSION].buckets
        .map((bucket) => bucket.key as string)
    .slice(0, size),
        composite: sortBy(
        flatten(
          aggregations[SERVICE_FRAMEWORK_NAME].buckets.map((bucket) =>
            bucket[SERVICE_FRAMEWORK_VERSION].buckets.map((versionBucket) => ({
              doc_count: versionBucket.doc_count,
              name: toComposite(bucket.key, versionBucket.key),
            }))
          )
        ),
        'doc_count'
      )
        .reverse()
        .slice(0, size)
        .map((composite) => composite.name),
    },
    language: {
      name: aggregations[SERVICE_LANGUAGE_NAME].buckets
        .map((bucket) => bucket.key as string)
    .slice(0, size),
        version: aggregations[SERVICE_LANGUAGE_VERSION].buckets
        .map((bucket) => bucket.key as string)
    .slice(0, size),
        composite: sortBy(
        flatten(
          aggregations[SERVICE_LANGUAGE_NAME].buckets.map((bucket) =>
            bucket[SERVICE_LANGUAGE_VERSION].buckets.map((versionBucket) => ({
              doc_count: versionBucket.doc_count,
              name: toComposite(bucket.key, versionBucket.key),
            }))
          )
        ),
        'doc_count'
      )
        .reverse()
        .slice(0, size)
        .map((composite) => composite.name),
    },
    runtime: {
      name: aggregations[SERVICE_RUNTIME_NAME].buckets
        .map((bucket) => bucket.key as string)
    .slice(0, size),
        version: aggregations[SERVICE_RUNTIME_VERSION].buckets
        .map((bucket) => bucket.key as string)
    .slice(0, size),
        composite: sortBy(
        flatten(
          aggregations[SERVICE_RUNTIME_NAME].buckets.map((bucket) =>
            bucket[SERVICE_RUNTIME_VERSION].buckets.map((versionBucket) => ({
              doc_count: versionBucket.doc_count,
              name: toComposite(bucket.key, versionBucket.key),
            }))
          )
        ),
        'doc_count'
      )
        .reverse()
        .slice(0, size)
        .map((composite) => composite.name),
    },
  },
}
```